[WebGPU] Support memoryless textures for render attachments

mwyrzykowski opened the pull request:
#### 02b481e73663f84fbd02e067238a4993cd8d7a00
<pre>
[WebGPU] Support memoryless textures for render attachments
<a href="https://bugs.webkit.org/show_bug.cgi?id=250064">https://bugs.webkit.org/show_bug.cgi?id=250064</a>
&lt;radar://103862028&gt;

Reviewed by NOBODY (OOPS!).

Automatically support memoryless render target attachments, e.g., for
MSAA targets and depth buffers, when the attachment is only used
as a render target and not used in subsequent passes.

This would for example reduce the memory usage of a fullscreen
WebGPU web app on iPhone 14 Pro from ~128MB to ~14MB, assuming
4x MSAA and 32bpp render+depth targets, accounting only for the MSAA,
depth, and resolve targets.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::storeAction):
Default to don&apos;t care for memoryless textures to avoid recreating for example
a depth buffer which is specified with store but only used in one render pass.

(WebGPU::addTextureStorageIfNeeded):
(WebGPU::CommandEncoder::beginRenderPass):
Recreate textures without memoryless storage if needed.

* Source/WebGPU/WebGPU/Texture.h:
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::storageMode):
Extract common logic to function.

(WebGPU::createTextureDescriptor):
Extract common logic to function.

(WebGPU::Device::createTexture):
Call extracted function.

(WebGPU::Texture::recreateBackingWithStorage):
Recreate texture without memoryless storage.

(WebGPU::Texture::createView):
Don&apos;t create a view when the format is identical as this is not
compatible with memoryless textures.

* Source/WebGPU/WebGPU/TextureView.h:
(WebGPU::TextureView::create):
Cache a reference to the Texture instance.

* Source/WebGPU/WebGPU/TextureView.mm:
(WebGPU::TextureView::TextureView):
(WebGPU::TextureView::recreateBackingWithStorageIfNeeded):
Recreate the texture and texture view if needed.

* Websites/webkit.org/demos/webgpu/scripts/two-pass-instanced-textured-cube.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/two-pass-instanced-textured-cube.html: Added.
Add a sample that illustrates how memoryless textures, initially used, will no longer
be used due to inefficient construction of render passes.

* Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:
Add sample to show how to opt-in to memoryless textures.

* Source/WebGPU/WebGPU/HardwareCapabilities.h:
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::apple3):
(WebGPU::mergeBaseCapabilities):
Add a flag for whether or not memoryless textures are supported.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/02b481e73663f84fbd02e067238a4993cd8d7a00

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103265 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12389 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36246 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112509 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13419 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3296 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95490 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/110773 "Built successfully") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109039 "Build is being retried. Recent messages:Running configure-build; Cleaned up git repository; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed webkitperl tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10305 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/93437 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37937 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92122 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24983 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79664 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5786 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26388 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5961 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2902 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11945 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45894 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7711 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebGPU/WebGPU/TextureView.h:
```diff
@@ -61,17 +61,19 @@ class TextureView : public WGPUTextureViewImpl, public RefCounted<TextureView> {
     const std::optional<WGPUExtent3D> m_renderExtent;
 
     const Ref<Device> m_device;
+    Texture* m_textureBacking { nullptr };
```
Yikes, raw pointer? Usually raw pointers indicate we haven't thought hard enough about lifetimes.

mwyrzykowski replied:
oh I think I can use Ref<Texture> here instead

litherum commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -129,6 +129,12 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
     return true;
 }
 
+static void addTextureStorageIfNeeded(TextureView& texture, WGPUStoreOp storeOp, WGPULoadOp loadOp)
```
nit: "add" implies that the storage will grow every time the function is called. Maybe "populate" or "inflate" or "allocate"?

litherum commented on Websites/webkit.org/demos/webgpu/two-pass-instanced-textured-cube.html:
```diff
@@ -0,0 +1,39 @@
+<!DOCTYPE html>
+<html>
```
How far away are we from using real tests for our work? I'm getting a bit antsy with all these custom demos, and no real tests.

mwyrzykowski replied:
Tests are very high priority! I think ref-tests were suggested possibly? I filed a bug for unit tests. While important, I don't think unit tests would go far in helping is to actually validate functionally things are working.

I just need to figure out how to setup ref-tests or any other suggested testing method - suggestions / pointers are welcomed

litherum replied:
I think we should use _only_ the CTS for testing. If there's a test that we want to write that isn't in the CTS, we should write it and upstream it.

mwyrzykowski replied:
this is the CTS right? https://gpuweb.github.io/cts/standalone/?q=webgpu:*

mwyrzykowski replied:
I don't think we can get to a state where we can only use CTS tests or only upstream CTS tests until we have made more progress on the compiler, since I don't think we can submit a CTS test that has metal source

litherum commented on Source/WebGPU/WebGPU/TextureView.mm:
```diff
@@ -51,6 +53,16 @@
 
+void TextureView::recreateBackingWithStorageIfNeeded()
+{
+    if (m_texture.storageMode != MTLStorageModeMemoryless)
```
Are you sure we don't want to shadow this state ourselves?

mwyrzykowski replied:
I don't think there is any benefit to shadowing it and it is not clear from the specification that Metal is not allowed to change the storageMode internally so we could make a member function in `Texture` that does something like:
```
MTLStorageMode Texture::storageMode()
{
  return m_texture.storgageMode;
}
```
but that seems unnecessary at this point

litherum commented on Source/WebGPU/WebGPU/TextureView.mm:
```diff
@@ -51,6 +53,16 @@
     m_texture.label = label;
 }
 
+void TextureView::recreateBackingWithStorageIfNeeded()
```
Because this function doesn't alter the programming model of this class, can we make this function `const`, and make the data member `mutable`? (Do you think that's actually worse, or better?)

mwyrzykowski replied:
I think that is what I did initially and then I reverted the mutable, it seemed about the same from my perspective

mwyrzykowski replied:
I will go back to mutable, it makes more sense as I am fixing the aliasing issue in TextureView

litherum commented on Source/WebGPU/WebGPU/TextureView.mm:
```diff
@@ -51,6 +53,16 @@
+
+    ASSERT(m_textureBacking);
+    m_textureBacking->recreateBackingWithStorage();
+    m_texture = m_textureBacking->texture();
```
I think this is wrong. `TextureView::m_texture` doesn't point to the same object as a `Texture::m_texture`. The view is different than the thing it's viewing.

In Metal, there is no distinct type that means "texture view". Indeed; the type of a texture view in Metal is just `MTLTexture`. That's what `TextureView::m_texture` points to: a texture view, not the original texture.

Maybe we should `using MTLTextureView = MTLTexture;` in our own code, to make it a bit more clear where views are used vs where non-views are used.

If we allocate storage for a WebGPU texture, then we're going to have to find all the derived TextureView objects, and re-create all their texture views. And, since bind groups hold references to texture views, we'll have to find all the bind groups that point to all those texture views, and recreate them.

mwyrzykowski replied:
In order to support memoryless textures, `TextureView::m_texture` should point to the same object as `Texture::m_texture` and this is what this patch does. 

We don't need to worry about bind groups holding references to memoryless textures because memoryless is only created when `descriptor.usage == WGPUTextureUsage_RenderAttachment` and to be in a bind group, the resource needs `(descriptor.usage & WGPUTextureUsage_TextureBinding)` 

There is an issue as you point out with multiple render target TextureView instances referencing the same texture. 

If you create multiple TextureView instances from the same render target, some with storeOp=store and some with storeOp=discard, then some of the TextureView instances with reference the memoryless texture and others will reference the memory-backed texture. That seems wrong since the backing Texture instance will no longer have a reference to the memoryless texture.

I'm not sure if it would cause incorrect behavior but I'll fix that issue because it seems at least it would be confusing

litherum commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -129,6 +129,12 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
 
+static void addTextureStorageIfNeeded(TextureView& texture, WGPUStoreOp storeOp, WGPULoadOp loadOp)
+{
+    if (storeOp == WGPUStoreOp_Store || loadOp == WGPULoadOp_Load)
```
How often do you expect this to be true?

mwyrzykowski replied:
It is hard to say how web developers will use WebGPU, but if logically they specify a store depth operation because the color attachment has a store operation, they will unfortunately hit this path even if they never read from their depth texture.

It only applies to textures which only have `MTLTextureUsageRenderTarget` and not `MTLTextureUsageShader*` so if they use this render target as an attachment in a forward renderer multipass algorithm, then they will never hit this path. Or IOSurface-backed textures will never hit this path, they are always memory backed

mwyrzykowski replied:
oh but this check is missing the memoryless part, that happens inside `recreateBackingWithStorageIfNeeded`, so `recreateBackingWithStorageIfNeeded` may get called unnecessary but it would immediately return. I'm not sure if that is a concern

litherum commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -152,26 +158,28 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
             attachment.clearColor.b,
             attachment.clearColor.a);
 
+        addTextureStorageIfNeeded(fromAPI(attachment.view), attachment.resolveTarget ? WGPUStoreOp_Discard : attachment.storeOp, attachment.loadOp);
```
We have to be really careful here. It's possible to submit commands in a different order than they were recorded. This code executes on the recording timeline, not the submission timeline. Naively, the logic about "X happens before Y" needs to be done on the submission timeline, because that's the order the GPU sees them.

I think the current patch's implementation of `addTextureStorageIfNeeded()` coincidentally solves this: Imagine there's a command buffer that is recorded using the memoryless texture, but the texture gets backing store allocated (and maybe even used!) in the window of time between our command buffer getting created and submitted. If our command buffer neither reads nor writes to the texture (which is the current logic of `addTextureStorageIfNeeded()`), then it's sort of a benign unobservable error that we recorded the command buffer referencing the wrong Metal texture.

On the other hand, the optimization only works if content _neither reads nor writes_ to the texture, which seems like a pattern I wouldn't expect to be common; the application could just ... not bind the render target at all, and have the same behavior.

If we tried to make the optimization more broadly applicable, and modified `addTextureStorageIfNeeded()` to only consider the store op, but not the load op, then that's incorrect because of the reordering problem described above.

mwyrzykowski replied:
Exactly. The optimization is primarily for MSAA where you need the samples but can resolve the 4x (for example) samples -> 1x prior to committing any stores. Also with MSAA, depth sample count must match color sample count, so it is possible to eliminate the 4x 32bpp depth buffer memory too.

A non-MSAA optimization is for single pass rendering: you can eliminate the storage of the depth buffer due to how TBDR work. A depth attachment is still needed however but it seems like it's mostly for validation.

litherum commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -152,26 +158,28 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
         mtlAttachment.texture = fromAPI(attachment.view).texture();
         mtlAttachment.level = 0;
         mtlAttachment.slice = 0;
         mtlAttachment.depthPlane = 0;
         mtlAttachment.loadAction = loadAction(attachment.loadOp);
-        mtlAttachment.storeAction = attachment.resolveTarget ? MTLStoreActionStoreAndMultisampleResolve : storeAction(attachment.storeOp);
 
         if (attachment.resolveTarget) {
-            mtlDescriptor.colorAttachments[i].resolveTexture = fromAPI(attachment.resolveTarget).texture();
+            mtlAttachment.resolveTexture = fromAPI(attachment.resolveTarget).texture();
```
Is this correct? This seems like the sort of thing that should be covered by a test. If nothing in the CTS covers this, can we write a test and upstream it to the CTS?

mwyrzykowski replied:
`mtlAttachment` and `mtlDescriptor.colorAttachments[i]` are the same, there is no logical change here

litherum commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -2417,17 +2432,26 @@ static WGPUExtent3D computeRenderExtent(const WGPUExtent3D& baseSize, uint32_t m
+    id<MTLTexture> texture;
+
+    if (m_texture.pixelFormat == pixelFormat && m_texture.textureType == textureType && m_texture.mipmapLevelCount == levels.length && m_texture.arrayLength == slices.length && !levels.location && !slices.location)
+        texture = m_texture;
```
How confident are we that `newTextureViewWithPixelFormat` doesn't do this under the hood?

mwyrzykowski replied:
`newTextureViewWithPixelFormat` returns `nil` for memoryless textures, that is why this code is needed

mwyrzykowski replied:
To limit the scope of this change I can also only invoke this path when `m_texture.storageMode == MTLStorageModeMemoryless`

litherum commented on Source/WebGPU/WebGPU/Texture.mm:
```diff
@@ -2417,17 +2432,26 @@ static WGPUExtent3D computeRenderExtent(const WGPUExtent3D& baseSize, uint32_t m
-    id<MTLTexture> texture = [m_texture newTextureViewWithPixelFormat:pixelFormat textureType:textureType levels:levels slices:slices];
-    if (!texture)
-        return TextureView::createInvalid(m_device);
+    id<MTLTexture> texture;
```
What happens when you bind the texture to a bind group? I think we'd need to have affordances for that (probably by just unconditionally giving it backing storage immediately?)

mwyrzykowski replied:
Memoryless textures are only created with `descriptor.usage == WGPUTextureUsage_RenderAttachment`, so they are not usable in bind groups.

mwyrzykowski committed:
[WebGPU] Support memoryless textures for render attachments
https://bugs.webkit.org/show_bug.cgi?id=250064
<radar://103862028>

Reviewed by NOBODY (OOPS!).

Automatically support memoryless render target attachments, e.g., for
MSAA targets and depth buffers, when the attachment is only used
as a render target and not used in subsequent passes.

This would for example reduce the memory usage of a fullscreen
WebGPU web app on iPhone 14 Pro from ~128MB to ~14MB, assuming
4x MSAA and 32bpp render+depth targets, accounting only for the MSAA,
depth, and resolve targets.

* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::storeAction):
Default to don't care for memoryless textures to avoid recreating for example
a depth buffer which is specified with store but only used in one render pass.

(WebGPU::addTextureStorageIfNeeded):
(WebGPU::CommandEncoder::beginRenderPass):
Recreate textures without memoryless storage if needed.

* Source/WebGPU/WebGPU/Texture.h:
* Source/WebGPU/WebGPU/Texture.mm:
(WebGPU::storageMode):
Extract common logic to function.

(WebGPU::createTextureDescriptor):
Extract common logic to function.

(WebGPU::Device::createTexture):
Call extracted function.

(WebGPU::Texture::recreateBackingWithStorage):
Recreate texture without memoryless storage.

(WebGPU::Texture::createView):
Don't create a view when the format is identical as this is not
compatible with memoryless textures.

* Source/WebGPU/WebGPU/TextureView.h:
(WebGPU::TextureView::create):
Cache a reference to the Texture instance.

* Source/WebGPU/WebGPU/TextureView.mm:
(WebGPU::TextureView::TextureView):
(WebGPU::TextureView::recreateBackingWithStorageIfNeeded):
Recreate the texture and texture view if needed.

* Websites/webkit.org/demos/webgpu/scripts/two-pass-instanced-textured-cube.js: Added.
(async helloCube.frameUpdate):
(async helloCube):
* Websites/webkit.org/demos/webgpu/two-pass-instanced-textured-cube.html: Added.
Add a sample that illustrates how memoryless textures, initially used, will no longer
be used due to inefficient construction of render passes.

* Websites/webkit.org/demos/webgpu/scripts/instanced-textured-cube.js:
Add sample to show how to opt-in to memoryless textures.

* Source/WebGPU/WebGPU/HardwareCapabilities.h:
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::apple3):
(WebGPU::mergeBaseCapabilities):
Add a flag for whether or not memoryless textures are supported.
